### PR TITLE
Remove various uses of raw pointers in the public API

### PR DESCRIPTION
**Pull request by randombit**
GH #3216

While some of these are changing the return types of non-deprecated functions, I think this is mostly OK; it's easy enough to use either variant as (for example) `std::unique_ptr<Public_Key> key(cert.subject_public_key())` and if it doesn't work at least the compiler gives a hint what is wrong (couldn't convert unique_ptr to raw ptr) vs the function just vanishing.

### CODE REVIEW

File: src/lib/x509/x509_obj.h [hunk 1]
```diff
@@ -80,15 +80,6 @@ class BOTAN_PUBLIC_API(2,0) X509_Object : public ASN1_Object
...
       bool check_signature(const Public_Key& key) const;
 
-      /**
-      * Check the signature on this data
-      * @param key the public key purportedly used to sign this data
-      *        the object will be deleted after use (this should have
-      *        been a std::unique_ptr<Public_Key>)
-      * @return true if the signature is valid, otherwise false
-      */
-      bool check_signature(const Public_Key* key) const;
```
Comment (reneme) [inline src/lib/x509/x509_obj.h:90]: Good riddance. 😌

**Review by reneme — APPROVED** [review]
Thanks a lot!

File: src/lib/x509/x509_obj.h (hunk 1 above)
Comment (randombit) [reply src/lib/x509/x509_obj.h:90]: Yeah a function that deletes its argument as a side effect of doing something else, wtf was I thinking.

File: src/lib/x509/x509cert.h [hunk 2]
```diff
@@ -38,15 +38,15 @@ class BOTAN_PUBLIC_API(2,0) X509_Certificate : public X509_Object
...
+      * certificate. If the public key bits was valid for X.509 encoding
+      * purposes but invalid algorithmically (for example, RSA with an even
+      * modulus) that will be detected at this point, and an exception will be
+      * thrown.
       *
-      * @return public key
+      * @return subject public key of this certificate
       */
-      Public_Key* subject_public_key() const;
+      std::unique_ptr<Public_Key> subject_public_key() const;
```
[the remark this answers is not in the document]
Comment (randombit) [reply src/lib/x509/x509cert.h:49]: Makes sense

**Pull request merged**